src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AbstractSpeculationAnalyzer.cs

```diff
@@ -380,6 +357,53 @@ private static bool SymbolsAreCompatibleCore(ISymbol symbol, ISymbol newSymbol,
             return false;
         }
 
+        private static bool CompareAcrossSemanticModels(ISymbol symbol, ISymbol newSymbol)
+        {
+            // SymbolEquivalenceComparer performs Location equality checks for locals, labels, range-variables and local
+            // functions. As we are comparing symbols from different semantic models, locations will differ. Hence
+            // perform minimal checks for these symbol kinds.
+
+            if (symbol.Kind != newSymbol.Kind)
+                return false;
+
+            if (symbol is ILocalSymbol localSymbol && newSymbol is ILocalSymbol newLocalSymbol)
+            {
+                return newSymbol.IsImplicitlyDeclared == symbol.IsImplicitlyDeclared &&
+                       symbol.Name == newSymbol.Name &&
+                       CompareAcrossSemanticModels(localSymbol.Type, newLocalSymbol.Type);
+            }
+
+            if (symbol is ILabelSymbol && newSymbol is ILabelSymbol)
+                return symbol.Name == newSymbol.Name;
+
+            if (symbol is IRangeVariableSymbol && newSymbol is IRangeVariableSymbol)
+                return symbol.Name == newSymbol.Name;
+
+            if (symbol is IParameterSymbol parameterSymbol &&
+                newSymbol is IParameterSymbol newParameterSymbol &&
+                parameterSymbol.ContainingSymbol.IsAnonymousOrLocalFunction() &&
+                newParameterSymbol.ContainingSymbol.IsAnonymousOrLocalFunction())
+            {
+                return symbol.Name == newSymbol.Name &&
+                       parameterSymbol.IsRefOrOut() == newParameterSymbol.IsRefOrOut() &&
+                       CompareAcrossSemanticModels(parameterSymbol.Type, newParameterSymbol.Type);
+            }
+
+            if (symbol is IMethodSymbol methodSymbol &&
+                newSymbol is IMethodSymbol newMethodSymbol &&
+                methodSymbol.IsLocalFunction() &&
+                newMethodSymbol.IsLocalFunction())
+            {
+                return symbol.Name == newSymbol.Name &&
+                       methodSymbol.Parameters.Length == newMethodSymbol.Parameters.Length &&
+                       CompareAcrossSemanticModels(methodSymbol.ReturnType, newMethodSymbol.ReturnType) &&
+                       methodSymbol.Parameters.Zip(newMethodSymbol.Parameters, (p1, p2) => (p1, p2)).All(
+                           t => CompareAcrossSemanticModels(t.p1, t.p2));
+            }
```

Comment:
❔ Can you clarify which of the bugs this change fixed? Can be a reply to this comment and not in code if there are already tests covering regressions.

Reply:
this was part of the https://github.com/dotnet/roslyn/issues/38347.  we need to check the previous and current local function. T his means checking the parameters of it.  And we need to ensure that all those checks work across semantic models.